### ISSUE TITLE
target-cpu=native should also check all the target-features

### ISSUE THREAD

**Issue by GabrielMajeri** [rust-lang/rust#54688]
Because LLVM doesn't know what `target-cpu=native` is, we currently [convert this to an actual CPU family name](https://github.com/rust-lang/rust/commit/1fd45a13dee17701fc0aeaa847c1919d485d09fd), relying on the [`llvm::sys::getHostCPUName`](http://llvm.org/doxygen/namespacellvm_1_1sys.html#a2e8cdc0e591685c9156af3d0d4fdae06) function. So if the current CPU's family is Haswell, we convert `target-cpu=native` to `target-cpu=haswell`.

As @EFanZh noticed [here](https://github.com/rust-lang/rust/issues/38218#issuecomment-424210340) and [here](https://github.com/rust-lang/rust/issues/48464), some Intel Pentiums belong to the Haswell microarch, but they lack AVX.

<hr/>

Here are the flags Clang generates for `-march=native`:

```sh
$ clang -E - -march=native -###
-target-cpu skylake -target-feature +sse2 -target-feature +cx16 -target-feature +sahf
...
-target-feature -avx512ifma -target-feature -avx512dq
```

Not only does it set the right `target-cpu` for optimization purposes, it actually manually enables / disables all the features available / not available on the host. This ensures no matter what the microarch is, only available features will be used.

<hr/>

We should probably use the [`llvm::sys::getHostCPUFeatures`](http://llvm.org/doxygen/namespacellvm_1_1sys.html#a4cdccbe62562aeff32fb137a042d5432) to manually get a list of supported target features. This is [what Clang does](https://github.com/llvm-mirror/clang/blob/024b9a4311cfd57ecb7f29d1138dc975fede4752/lib/Driver/ToolChains/Arch/X86.cpp#L116).

I think this issue will supersede both rust-lang/rust#38218 and rust-lang/rust#48464. If we can fix this, we will never generate code with unavailable target features.

Reproducing these errors is pretty hard, unless you have a machine with a Haswell Intel Pentium. 
I think it might be possible to reproduce this in QEMU with `-cpu Haswell,-avx`

**Comment by gnzlbg** [comment]
> We should probably use the llvm::sys::getHostCPUFeatures to manually get a list of supported target features. This is what Clang does.

Alternatively, we could also use the `is_{arch}_feature_detected!` macros to do this. I'll work on this.

**Comment by gendx** [comment]
To add another example, `target-cpu=native` doesn't detect the `aes` feature on broadwell (see https://github.com/rust-lang/rust/issues/67836).

**Comment by as-com** [comment]
I submitted a PR for a fix: #80749
